### ISSUE TITLE
Issues not tracking although `php artisan sentry:test` works fine

### ISSUE THREAD

**Issue by johnjrmurray** [getsentry/sentry-laravel#328]
I'm having a strange issue on a fresh install of Laravel on getting Sentry working. We have 2 projects on different versions on Laravel & Sentry-Laravel that are working perfectly fine on the same PHP version and server. Never had an issue with the install until today.

PHP 7.2.10
Laravel Framework 6.15.1
sentry/sentry-laravel 1.6.1
OS Red Hat Enterprise Linux Server release 7.7

This is for the  /debug-sentry route. Exception appears on screen but nothing is logged in Sentry. As mentioned in the title `php artisan sentry:test` works fine and the 

Definitely getting into the exception handler as a dd(app()->bound('sentry')) within \App\Exceptions\Handler::report returns a `true`.

Worth mentioning we are behind a corporate proxy, the http_proxy config is set in config/sentry.php and in the .env file.

Any advice on how to debug this would be great.

Thanks

**Comment by stayallive** [comment]
Well this is a really interesting one... `php artisan sentry:test` uses the same container instance as the web requests, there are 2 things that I can think of you could quickly try/check.

- Config caching (`php artisan config:clear`)? (although his should also affect artisan)
- PHP-FPM process wasn't restarted/reloaded after config/code changes

If the artisan command is working I think it's safe to assume the DSN & proxy config is working as expected.

```php
    if (app()->bound('sentry') && $this->shouldReport($exception)) {
        app('sentry')->captureException($exception);
    } 
```

We recommend the above snippet, you could check the exception is not ignored by the `$this->shouldReport($exception)` guard.

The last thing I can think of is that you are using `die` or `exit` (when you were not debugging) on the `/debug-sentry` route or some other way to cut the request short not allowing Sentry to flush it's internal queue (it tries to do that as late as possible). You can validate this by adding `\Sentry\Laravel\Integration::flushEvents();` after the `captureException` call.

**Comment by johnjrmurray** [comment]
Thanks for the response.

I've tried moving the below out of the conditional, just to be sure 
`app('sentry')->captureException($exception);`

Tried clearing the config  and rebooted the server for good measure.

Interesting observation when introducing ``\Sentry\Laravel\Integration::flushEvents();` into the report block, this causes the page to time out on load.

**Comment by stayallive** [comment]
Hmm that makes me thing there might be something wrong with the proxy and/or DSN config after all, still doesn't compute with me why `artisan sentry:test` does work correctly...

Unfortunately we are still hard at work trying to correctly implement a logger so we can easily observe any transport errors (getsentry/sentry-php#907).

So the only thing you can do is place some testing around this part in your vendor folder (without re-creating the transport which is also an option: https://docs.sentry.io/platforms/php/#transport) to figure out what exception might be thrown while trying to send your error: https://docs.sentry.io/platforms/php/#transport.

**Comment by johnjrmurray** [comment]
Looks like I've found where the problem lies, around vendor/sentry/sentry/src/HttpClient/HttpClientFactory.php:102:

`        if (null === $httpClient && null !== $options->getHttpProxy()) {
            if (class_exists(GuzzleHttpClient::class)) {
                /** @psalm-suppress InvalidPropertyAssignmentValue */
                $this->httpClient = GuzzleHttpClient::createWithConfig([
                    GuzzleHttpClientOptions::PROXY => $options->getHttpProxy(),
                ]);
            } elseif (class_exists(CurlHttpClient::class)) {
                /** @psalm-suppress InvalidPropertyAssignmentValue */
                $this->httpClient = new CurlHttpClient($this->responseFactory, $this->streamFactory, [
                    CURLOPT_PROXY => $options->getHttpProxy(),
                ]);
            } else {
                throw new \RuntimeException('The "http_proxy" option requires either the "php-http/curl-client" or the "php-http/guzzle6-adapter" package to be installed.');
            }
        }

        if (null === $httpClient) {
            $httpClient = HttpAsyncClientDiscovery::find();
        }`

Proxy config is set correctly within the GuzzleHttpClient block to `this->httpClient`, but is then overwritten as $httpClient is still null at this point.

The following change fixes the issue:
`        if (null === $httpClient && null !== $options->getHttpProxy()) {
            if (class_exists(GuzzleHttpClient::class)) {
                /** @psalm-suppress InvalidPropertyAssignmentValue */
                $this->httpClient = GuzzleHttpClient::createWithConfig([
                    GuzzleHttpClientOptions::PROXY => $options->getHttpProxy(),
                ]);
            } elseif (class_exists(CurlHttpClient::class)) {
                /** @psalm-suppress InvalidPropertyAssignmentValue */
                $this->httpClient = new CurlHttpClient($this->responseFactory, $this->streamFactory, [
                    CURLOPT_PROXY => $options->getHttpProxy(),
                ]);
            } else {
                throw new \RuntimeException('The "http_proxy" option requires either the "php-http/curl-client" or the "php-http/guzzle6-adapter" package to be installed.');
            }
        }

        $httpClient = $this->httpClient;

        if (null === $httpClient) {
            $httpClient = HttpAsyncClientDiscovery::find();
        }`

**Comment by stayallive** [comment]
Interesting... would you be willing to submit this as a PR to https://github.com/getsentry/sentry-php?

Maybe this version would be even better / more logical?

```php
        if (null === $httpClient && null !== $options->getHttpProxy()) {
            if (class_exists(GuzzleHttpClient::class)) {
                /** @psalm-suppress InvalidPropertyAssignmentValue */
                $httpClient = GuzzleHttpClient::createWithConfig([
                    GuzzleHttpClientOptions::PROXY => $options->getHttpProxy(),
                ]);
            } elseif (class_exists(CurlHttpClient::class)) {
                /** @psalm-suppress InvalidPropertyAssignmentValue */
                $httpClient = new CurlHttpClient($this->responseFactory, $this->streamFactory, [
                    CURLOPT_PROXY => $options->getHttpProxy(),
                ]);
            } else {
                throw new \RuntimeException('The "http_proxy" option requires either the "php-http/curl-client" or the "php-http/guzzle6-adapter" package to be installed.');
            }
        }

        if (null === $httpClient) {
            $httpClient = HttpAsyncClientDiscovery::find();
        }
```

I'd like you to be able to take the credit for this bugfix, if you don't want to or have the time let me know so I can do it on your behalf 👍

**Comment by iSWORD** [comment]
Is this merged into the Laravel integration in/before 1.7.0? I still cannot get any reports except from `php artisan sentry:test`. This is affecting most of our production apps.

**Comment by johnjrmurray** [comment]
a composer update to sentry/sentry (the SDK) should do the trick.

**Comment by stayallive** [comment]
This issue was fixed in [`sentry/sentry:2.3.2`](https://github.com/getsentry/sentry-php/releases/tag/2.3.2) (`sentry/sentry` is a dependency of `sentry/sentry-laravel`) so as long as you have updated to that version (run `composer info sentry/sentry` to check the current version and `composer update` if you are not on `2.3.2`) the `http_proxy` option should function correctly.

**Comment by iSWORD** [comment]
Thank you so much for your prompt responses @johnjrmurray and @stayallive 

As instructed, I have opened a new issue at https://github.com/getsentry/sentry-laravel/issues/336 since I confirmed the version of sentry/sentry to be 2.3.2.
